lib/ide/ide/src/controller.rs

```diff
@@ -0,0 +1,377 @@
+                    Ok(read_future.await?),
+                Err(e) =>
+                    Err(e.into()),
+            }
+        }
+
+        /// Creates a new module controller.
+        pub async fn create_module_controller
+        (&self, loc:&module::Location) -> FallibleResult<module::StrongHandle> {
+            let data = module::Data{
+                path     : loc.clone(),
+                contents : self.read_module(loc.clone()).await?,
+                parent   : self.clone(),
+            };
+            let module = self.insert_module(data).await?;
+            Ok(module)
+        }
+    }
+    ////
+
+    impl WeakHandle {
+        /// Returns a module controller for given module location.
+        ///
+        /// Reuses existing controller if possible.
+        /// Creates a new controller if needed.
+        pub async fn open_module(&self, loc:&module::Location) -> FallibleResult<module::StrongHandle> {
+            if let Some(existing_controller) = self.lookup_module(loc.clone()).await? {
+                Ok(existing_controller)
+            } else {
+                Ok(self.create_module_controller(&loc).await?)
+            }
+        }
+
+        pub async fn open_file_text(&self, _path:&std::path::Path) -> FallibleResult<text::StrongHandle> {
+            // TODO [mwu] similar to the above
+            todo!()
+        }
+    }
+
+    /// boilerplate wrappers over stateful `Data` APIs
+    impl WeakHandle {
+        pub async fn lookup_module
+        (&self, loc:module::Location)
+        -> handle::Result<Option<module::StrongHandle>> {
+            self.with(move |data| data.lookup_module(&loc)).await
+        }
+        pub async fn insert_module
+        (&self, module:module::Data)
+        -> handle::Result<module::StrongHandle> {
+            self.with(|data| data.insert_module(module)).await
+        }
+        pub async fn call_fm<R>
+        (&self, f:impl FnOnce(&mut file_manager_client::Client) -> R)
+        -> handle::Result<R> {
+            self.with(|data| f(&mut data.file_manager)).await
+        }
+    }
+}
+
+/// Module controller.
+mod module {
+    use super::*;
+
+    pub enum FromParent {
+        FileChanged,
+    }
+    pub enum ToParent {
+
+    }
+
+
+    #[derive(Clone,Debug,Eq,Hash,PartialEq)]
+    pub struct Location(pub String);
+    impl Location {
+        pub fn to_path(&self) -> file_manager_client::Path {
+            let result = format!("./{}.luna", self.0);
+            file_manager_client::Path::new(result)
+        }
+    }
+
+    #[derive(Clone,Debug)]
+    pub struct Data {
+        pub path     : Location,
+        pub contents : String,
+        pub parent   : project::WeakHandle,
+    }
+
+    impl Data {
+        pub fn fetch_text(&self) -> impl Future<Output = FallibleResult<String>> {
+            let loc    = self.path.clone();
+            let parent = self.parent.clone();
+            async move {
+                parent.read_module(loc).await
```

Comment:
As you know, I don't like refering to the parent, mainly because now everyone who calls `fetch_text` should be prepared for an error "Parent does not exists" which never occurs.

I think the module controller should have an access to FileManagerClient and ask him for file content directly.